### PR TITLE
#62 configprovider rework

### PR DESCRIPTION
**Pull request by struberg**
move the complex and seldom used parts away from ConfigProvider.

### CODE REVIEW

File: api/src/main/java/org/eclipse/microprofile/config/spi/ConfigSource.java [hunk 1]
```diff
@@ -69,6 +79,6 @@
      *
      * @return the unique 'id' of the configuration source, e.g. 'property-file mylocation/myproperty.properties'
```
Comment (Emily-Jiang) [inline api/src/main/java/org/eclipse/microprofile/config/spi/ConfigSource.java]: The javadoc needs to be updated.

**Comment by struberg** [comment]
> I prefer not to change this class as it hides the ConfigBuilder, which is better not to be exposed too widely.

That exactly was the reason why I moved all those methods and classes from the ConfigProvider (user facing class) to the spi module and thus the integration layerl.

**Comment by Emily-Jiang** [comment]
In my view, making the ConfigBuilder a standalone classing is making it more visible rather than hiding it.

**Comment by struberg** [comment]
The point is to make ConfigProvider easier to grok while moving more complex use cases away from the cacual developer.

Btw, I need to fix the Copyright. It should not say "Copyright Mark Struberg and others", but "Copyright Romain Manni-Bucau and others".
Romain clearly wrote the initial version with a pull request to my original proposal on July 20th, 2016:
https://github.com/struberg/javaConfig/pull/2/commits/0945b23cbf9dadb75fb97551837e6ce3d26496f6
https://github.com/struberg/javaConfig/pull/2/commits/84112ecb8fc693ff55d0896b19dfa2a8a50b35ef

**Comment by OndrejM** [comment]
I see your point, @Emily-Jiang ,but I think ConfigBuilder really belongs to SPI rather than to the API package and I agree with @struberg. We could try hiding it even within the SPI package, e.g. into ConfigProviderResolver, but that's a little too unnatural.

**Comment by struberg** [comment]
> We could try hiding it even within the SPI package, e.g. into ConfigProviderResolver, 
> but that's a little too unnatural.

yes @OndrejM , that was my feeling as well. ConfigProviderResolver used to be an inner class of ConfigProvider (originally ConfigProvider#SPI). Moving the ConfigBuilder to this place seems a bit too much.

**Pull request merged**